Add tree editor example

JonasHelming opened the pull request:
Adds a new template `Tree Editor` to the generator which allows to generate an extension with a tree master detail editor.
Additionally, add a new menu Tree Editor with an entry to generate an example file to open in the editor.
The tree editor uses a coffee machine example model.

fixes #74

Co-authored-by: Florian Gareis <mail@zoker.me>
Signed-off-by: Lucas Koehler <lkoehler@eclipsesource.com>

marcdumais-work commented:
note: It's probably better to merge https://github.com/theia-ide/generator-theia-extension/pull/73 first. Then it's clear this is contributed under the new license.

JonasHelming commented:
I agree! The contributors are covered by existing agreements, though, they both implemented "on behalf of EclipseSource2

JonasHelming commented:
OK, #73 is merged by now

marcdumais-work commented:
The license change PR is merged - this repo is effectively if not officially re-licensed (see wiki for latest updates). Please rebase this PR to pick-up our latest master and along with it the new license and license headers.  Add any missing license headers on new files in this PR, whenever applicable. 

Thanks!

JonasHelming commented:
@lucas-koehler

lucas-koehler commented:
@JonasHelming @marcdumais-work I rebased the changes. No adaptions were necessary as all new files are part of the template.

marcdumais-work commented:
@vince-fugnitto would you have time to review this PR?

vince-fugnitto commented on templates/tree-editor/example-file/example-file-contribution.ts:
```diff
@@ -0,0 +1,43 @@
+            label: 'New Example File'
+        });
+    }
+}
```
newline.

vince-fugnitto commented on templates/tree-editor/tree/tree-editor-widget.ts:
```diff
@@ -0,0 +1,56 @@
+    export const WIDGET_ID = '<%= params.extensionPath %>-tree-editor';
+    export const EDITOR_ID = '<%= params.extensionPath %>.tree.editor';
+}
+    
```
Should be fixed.

vince-fugnitto replied:
I believe the previous comment still holds true.

vince-fugnitto commented on templates/tree-editor/tree/tree-model.ts:
```diff
@@ -0,0 +1,49 @@
+        ]
+    ]);
+
+}
```
newline.

vince-fugnitto commented on templates/tree-editor/tree/tree-schema.ts:
```diff
@@ -0,0 +1,464 @@
+    }
+  },
+  '$ref': '#/definitions/machine'
+};
```
newline.

vince-fugnitto reviewed (commented):
I reviewed quickly and wanted to know your opinion on a few matters:
- we should probably include a `readme` for the generated `tree-editor` example so users can understand the features, it was not clear from simply generating the extension.
- creating a `*.tree` (ex: `tree-example.tree`) file through the main-menu does not immediately open the file - this would make the extension easier to use especially in large workspaces.
- the `tree-editor` does not respect the `auto-save` feature, meaning if I were to update values in the editor the content is never saved unless users explicitly do so - should this not be present in the extension by default, else all extenders would need to fix it?
- the editor is not responsive, and there is no way to scroll on smaller displays - another bug each extender will need to address.

JonasHelming commented:
> I reviewed quickly and wanted to know your opinion on a few matters:
> 
> * we should probably include a `readme` for the generated `tree-editor` example so users can understand the features, it was not clear from simply generating the extension.
> * creating a `*.tree` (ex: `tree-example.tree`) file through the main-menu does not immediately open the file - this would make the extension easier to use especially in large workspaces.
> * the `tree-editor` does not respect the `auto-save` feature, meaning if I were to update values in the editor the content is never saved unless users explicitly do so - should this not be present in the extension by default, else all extenders would need to fix it?
> * the editor is not responsive, and there is no way to scroll on smaller displays - another bug each extender will need to address.

@vince-fugnitto : Very good feedback, thank you! We will address these issues and update the PR.

lucas-koehler commented:
@JonasHelming @vince-fugnitto I rebased the changes on the current master and added a commit to  address Vince's feedback. Thanks again for the great feedback :)

lucas-koehler commented:
@JonasHelming @vince-fugnitto I pushed a little fix for the new tree editor extension readme. Besides that I don't expect any further changes (except of course if there's additional feedback ;) )

vince-fugnitto commented on README.md:
```diff
@@ -32,6 +32,7 @@ Extension options
 1. `hello-world`: creates a simple extension which provides a command and menu item which displays a message.
 2. `widget`: creates the basis for a simple widget including a toggle command, alert message and button displaying a message.
 3. `labelprovider`: create a simple extension which adds a custom label (with icon) for `.my` files
+4. `tree-editor`: create a tree editor extension
```
nitpick: do you mind adding punctuation:

```suggestion
3. `labelprovider`: create a simple extension which adds a custom label (with icon) for `.my` files.
4. `tree-editor`: create a tree editor extension.
```

vince-fugnitto commented on templates/tree-editor/tree-frontend-module.ts:
```diff
@@ -0,0 +1,55 @@
+import { NewTreeExampleFileCommandContribution, NewTreeExampleFileMenuContribution } from './example-file/example-file-contribution';
+
+export default new ContainerModule(bind => {
+    // Bind Theia IDE contributions for the example file creation menu entry
```
nit:

```suggestion
    // Bind Theia IDE contributions for the example file creation menu entry.
```

vince-fugnitto commented on templates/tree-editor/tree-frontend-module.ts:
```diff
@@ -0,0 +1,55 @@
+    bind(CommandContribution).to(NewTreeExampleFileCommandContribution);
+    bind(MenuContribution).to(NewTreeExampleFileMenuContribution)
+
+    // Bind Theia IDE contributions for the tree editor
```
nit:

```suggestion
    // Bind Theia IDE contributions for the tree editor.
```

vince-fugnitto commented on templates/tree-editor/tree-frontend-module.ts:
```diff
@@ -0,0 +1,55 @@
+                TreeNodeFactory
+            );
+
+            // Bind options
```
nit:

```suggestion
            // Bind options.
```

vince-fugnitto commented on templates/tree-editor/tree-label-provider-contribution.ts:
```diff
@@ -0,0 +1,26 @@
+        return 'fa fa-coffee dark-purple';
+    }
+    
+    // We don't need to specify getName() nor getLongName() because the default uri label provider is responsible for them
```
nit:

```suggestion
    // We don't need to specify getName() nor getLongName() because the default uri label provider is responsible for them.
```

vince-fugnitto reviewed (commented):
@JonasHelming @lucas-koehler the changes look good to me, and my previous usability comments have been successfully addressed: https://github.com/theia-ide/generator-theia-extension/pull/75#issuecomment-725491412

I have some minor comments if you believe they should be fixed.

JonasHelming commented:
Thank you for the review, I addressed the comments!
@vince-fugnitto : I do not know whether you prefer us to directly squash or whether you want to see the changes in commits first, just let me know, I can squash anytime..

vince-fugnitto commented:
@JonasHelming yes please squash for a clean history :)

JonasHelming commented:
Done, as you previously did not want to be co author for typo suggestions, I did not add you. Please let me know, if you want to be added.

vince-fugnitto commented:
> Done, as you previously did not want to be co author for typo suggestions, I did not add you. Please let me know, if you want to be added.

No its fine :) I'm back from vacation so I believe it should be fine to merge. Do you need help with merging?

JonasHelming commented:
I cannot merge, so yes, I would appreciate "help with merging" :-)

JonasHelming commented:
A package update would also be wonderful

vince-fugnitto commented:
> A package update would also be wonderful

@JonasHelming I have published a latest release `v0.1.21`, can you confirm it works correctly?

JonasHelming commented:
looks good, thank you so much!
